.werft/util/werft.ts

```diff
@@ -0,0 +1,113 @@
+import { Span, Tracer, trace, context, SpanStatusCode } from '@opentelemetry/api';
+import { exec } from './shell';
+
+let werft: Werft;
+
+/**
+ * For backwards compatibility with existing code we expose a global Werft instance
+ */
+export function getGlobalWerftInstance() {
+    if (!werft) {
+        throw new Error("Trying to fetch global Werft instance but it hasn't been instantiated yet")
+    }
+    return werft
+}
+
+/**
+ * Class for producing Werft compatible log output and generating traces
+ */
+export class Werft {
+    private tracer: Tracer;
+    public rootSpan: Span;
+    private sliceSpans: { [slice: string]: Span } = {}
+    private currentPhaseSpan: Span;
+
+    constructor(job: string) {
+        if (werft) {
+            throw new Error("Only one Werft instance should be instantiated per job")
+        }
+        this.tracer = trace.getTracer("default");
+        this.rootSpan = this.tracer.startSpan(`job: ${job}`, { root: true, attributes: { 'werft.job.name': job } });
+
+        // Expose this instance as part of getGlobalWerftInstance
+        werft = this;
+    }
+
+    public phase(name, desc?: string) {
+        // When you start a new phase the previous phase is implicitly closed.
+        if (this.currentPhaseSpan) {
+            this.endPhase()
+        }
+
+        const rootSpanCtx = trace.setSpan(context.active(), this.rootSpan);
+        this.currentPhaseSpan = this.tracer.startSpan(`phase: ${name}`, {
+            attributes: {
+                'werft.phase.name': name,
+                'werft.phase.description': desc
+            }
+        }, rootSpanCtx)
+
+        console.log(`[${name}|PHASE] ${desc || name}`)
+    }
+
+    public log(slice, msg) {
+        if (!this.sliceSpans[slice]) {
+            const parentSpanCtx = trace.setSpan(context.active(), this.currentPhaseSpan);
+            const sliceSpan = this.tracer.startSpan(`slice: ${slice}`, undefined, parentSpanCtx)
+            this.sliceSpans[slice] = sliceSpan
+        }
+        console.log(`[${slice}] ${msg}`)
+    }
```

Comment:
Do I understand it correctly that this is adding a new span every time we log something to werft's console?

I think that is probably overkilling since we log stuff sporadically throughout our pipeline and most of the time they are not really useful inside the trace.

Reply:
We produce a single span per slice. Slices are implicitly created by Werft whenever you log something. From the spans that are generated I'd say that it is at the high verbosity level, otherwise we wouldn't have any of the spaces in the deploy phase in the screenshot below ([link to trace](https://ui.honeycomb.io/gitpod/datasets/werft/trace/uJQwGhCdSGX))
![Screenshot 2021-11-03 at 13 37 46](https://user-images.githubusercontent.com/83561/140061446-a04cc813-d955-4b41-b9e0-03a528bf81bf.png)